Remove the sympy_tokenize module in favor of the standard library tokenize

asmeurer opened the pull request:
<strike>I have not yet fixed the modifications to the tokenize module, which are

- factorial (x!)
- repeated decimals (0.[1])

I also need to add tests for sympify('α') in Python 3.</strike>

This is ready for review.

Fixes #4862.

asmeurer committed:
Remove the sympy_tokenize module in favor of the standard library tokenize

I have not yet fixed the modifications to the tokenize module, which are

- factorial (x!)
- repeated decimals (0.[1])

I also need to add tests for sympify('α') in Python 3.

Fixes #4862.

asmeurer committed:
Update the factorial_notation transformer to work with the standard library tokenize

Also add some more tests for it.

asmeurer committed:
Add repeated_decimal as a transformer on the standard library tokenize

Unfortunately, there is currently a regression, which is that .[1] doesn't
work. To handle this, we need to detect when . is an OP that cannot apply to
whatever is before it. Alternately, we could always parse . (as OP) + [INT] as
a repeated decimal, since it would never be valid syntax anyway.

asmeurer committed:
Fix sympify('.[1]') regression

asmeurer committed:
Add test for sympify(u'α') (which only works in Python 3)

asmeurer committed:
Use isdigit instead of isdecimal

The latter is not available in Python 2. While isdecimal is technically more
correct (isdigit will match certain Unicode characters), it doesn't matter
because they wouldn't be tokenized as NUMBER anyway.

asmeurer commented:
OK, this is ready for review.

asmeurer committed:
Add some more tests for repeated decimal

asmeurer commented:
Are there other new Python 3 syntax features aside from Unicode variable names that I should add tests for?

asmeurer commented:
I guess underscores in numeric literals. What else?

asmeurer committed:
Add a test for parsing numbers with underscores (Python 3.6+ only)

smichr commented on sympy/parsing/sympy_parser.py:
```diff
@@ -626,16 +620,91 @@ def convert_xor(tokens, local_dict, global_dict):
+
+        result.append((toknum, tokval))
+        if len(num) == 4:
+            # pre.post[repeatend] = a + b/c + d/e where a = pre, b/c = post,
```
repeatend -> repetend in this line and next

asmeurer replied:
Fixed, thanks.

asmeurer committed:
Fix spelling

asmeurer committed:
Add more repeated deciamal parsing tests, and remove an unused import

asmeurer committed:
Fix repeated decimals tokenizing when the repetend starts with 0

Python 2 tokenizes the first non-89 digits separately, and Python 3 tokenizes
the first 0 digits separately.

This has a minor regresssion, which is that sympify('.[0 1]') now parses the
same as sympify('.[01]'), because the tokenizer does not distinguish between
the two. This could be fixed by checking the column positions, but it doesn't
seem worth it.

Also:

- Checks for x are not necessary, because hex numbers cannot be floats.

- Fix some indentation.

- Make sure to clear num after it is added to result.

asmeurer committed:
Fix Sphinx docs for the parsing module

asmeurer committed:
Make docstring for repeated_decimals clearer

asmeurer commented:
So I not feel a little less strongly about using AST transformers. It turns out even the repeated decimals can't be done with them, because of the `.[12]` case, which isn't valid Python. 

Tokenizing is useful, because it works even for invalid Python, as long as whitespace isn't significant. The `tokenize` module seems quite generous in how it tokenizes things, even producing tokens for things that can never be valid Python (like `!`). 

We could actually backport Python 3 features to Python 2 using transformers, like Unicode names or matrix multiplication. I personally have no plans to do this, but if someone is interested it shouldn't be that hard.

asmeurer commented:
This is ready for full review now (the tests are passing).

asmeurer committed:
Fix comment

asmeurer commented on sympy/parsing/sympy_parser.py:
```diff
@@ -626,16 +620,98 @@ def convert_xor(tokens, local_dict, global_dict):
+            pre, post = num[0][1].split('.')
+            repetend = num[2][1]
+            if len(num) == 5:
+                repetend += num[3][1]
```
In case anyone is wondering why the joining is done here instead of [above](https://github.com/sympy/sympy/pull/14085/files#diff-c3f3cb4004234277039032a7201a70a8R644) it's because we still have to account for truncating `result` [here](https://github.com/sympy/sympy/pull/14085/files#diff-c3f3cb4004234277039032a7201a70a8R665). If we simply join and keep `len(num) == 4`, the size of `result` will be off.

smichr commented:
Should the underscore separator be handled in the repetend?

```
decimal ok with either
>>> S('.123 456')
0.123456000000000
>>> S('.123_456')
0.123456000000000

repetend ok with space
>>> S('.12[345 123]')
1371679/11111100

...but not with underscore
>>> S('.12[345_123]')
Traceback (most recent call last):
...
TypeError: 'Float' object does not support indexing
```

asmeurer committed:
Add support for underscores in repentends in Python 3.6+ (S('1.2[3_4]'))

This only works in Python 3.6+ because only there are numeric literals with
underscores tokenized as such.

asmeurer commented:
The spaces in the repetend are supported accidentally. See the code comments and commit messages. The issue is that Python parses 0123456789 is two separate tokens, since it isn't a valid integer literal (in different ways depending on if it is Python 2 or Python 3). To support this, we join concurrent integer tokens in the repetend. This also means that any concurrent tokens with space are also allowed, but it's only because didn't feel it was worth it to disallow it. It would require a lot of logic for very little gain. 

Underscores are only supported in Python 3.6. Adding support for them there is quite easy. I've pushed up a commit that does it. It only works in Python 3.6+ because only there are underscores in numeric literals tokenized as such.  As I said in a previous comment, I don't really have much interest here in trying to backport language features so that things work uniformly across Python versions, although this is possible to do, at least for most cases, if anyone wants to try it (we could have a separate set of `language_backport_transformers`).

asmeurer commented:
I didn't realize underscores are allowed in float literals as well. I've fixed that and added a test.

asmeurer committed:
Fix underscores in the float part in repeated decimal parsing

smichr commented:
Looks simpler than what I would have done.

+1
